### PR TITLE
Class and parse_crf for 1D data 

### LINKED ISSUE #301: Add ability to add 1D emissions timeseries (e.g. global total)
Want to allow emissions timeseries data to be added to the object store and plotted. May want to set this up as a new class or something based on current classes.

### PR DESCRIPTION
**Pull request by SutarPrasad**
* **Summary of changes** (Bug fix, feature, docs update, ...)

- [x] Created 1D class for the OneDTtimeseries data
- [x] Added standardised function
- [x] Added parse_crf function
- [x] Added OneDtimeseries inside specifications, enum and created load_oned-parser
- [x] Added Dataschema validation method

* **Please check if the PR fulfills these requirements**

- [x] Closes #301 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://docs.openghg.org/development/python_devel.html#testing) if fixing a bug or adding a new feature
- [x] All code checks passing - `black --line-length 110` run over code, `mypy` and `flake8` not showing any errors
- [ ] Documentation and tutorials updated/added
- [x] Added an entry in the latest `CHANGELOG.md` file if fixing a bug or adding a new feature
- [x] Added any new requirements to `requirements.txt` and `recipes/meta.yaml`

### CODE REVIEW

**Comment by brendan-m-murphy** [comment]
I know we've been calling it "1D timeseries" for a while, but time is 1 dimensional, so we could probably just call it "timeseries". If you have spatial coordinates as well, I guess it is called "spatio-temporal" data sometimes. (If 1D time series just means 1 data variable, then I would probably call that "time series" as well; multiple variables with a time dimension is usually called a "multiple time series")

File: openghg/standardise/oned_timeseries/_crf.py
Comment (gareth-j) [inline openghg/standardise/oned_timeseries/_crf.py:10]: Do we always want a default domain?

File: openghg/standardise/oned_timeseries/_crf.py
Comment (gareth-j) [inline openghg/standardise/oned_timeseries/_crf.py:21]: Remove the types from the docstring

File: openghg/standardise/oned_timeseries/_crf.py
Comment (gareth-j) [inline openghg/standardise/oned_timeseries/_crf.py:33]: Move this import to the top of the file

File: openghg/store/_oneD_timeseries.py
Comment (gareth-j) [inline openghg/store/_oneD_timeseries.py:202]: Why do we convert to a `tuple` here?

File: openghg/store/spec/_specification.py
Comment (gareth-j) [inline openghg/store/spec/_specification.py:18]: Maybe `oned_timeseries` ?

File: openghg/store/spec/_specification.py
Comment (gareth-j) [inline openghg/store/spec/_specification.py:46]:
```suggestion
        "oned_timeseries": OneDTimeseries,
```

File: openghg/types/_enum.py
Comment (gareth-j) [inline openghg/types/_enum.py:64]:
```suggestion
class OneDTypes(Enum):
```

We should stick to CamelCase

**Review by gareth-j — CHANGES_REQUESTED** [review]
Looks good, just need some changes and then the tests adding.

**Comment by rt17603** [comment]
> I know we've been calling it "1D timeseries" for a while, but time is 1 dimensional, so we could probably just call it "timeseries". If you have spatial coordinates as well, I guess it is called "spatio-temporal" data sometimes. (If 1D time series just means 1 data variable, then I would probably call that "time series" as well; multiple variables with a time dimension is usually called a "multiple time series")

I would agree with this with an additional suggestion of making this `FluxTimeseries` if this is specifically for example of flux/emissions data as a timeseries data or a general `Timeseries` object if this was being made as suitable for any timeseries (in the second instance though we may want to think about how to relate ObsSurface to this as well).

File: openghg/standardise/_standardise.py
Comment (rt17603) [reply openghg/standardise/_standardise.py:849]: The use of `domain` is an interesting one because in this case we're not using it to describe an explicit lat-lon range like we are for the 3D emissions and footprints.

Having a tag for the area we're considering is important but I wonder if we could:
 - add a `region` keyword as the descriptor for the area the flux data represents e.g. `region="uk"`
 - keep `domain` as optional but only use this in cases where this represents a sum over one of our named domains e.g. "europe" (not used often I'd image but perhaps useful to have)

File: openghg/store/_oneD_timeseries.py
Comment (rt17603) [inline openghg/store/_oneD_timeseries.py:165]: Should match to whatever alternative `data_type` name is chosen (e.g. "flux_timeseries")

File: openghg/standardise/oned_timeseries/_crf.py
Comment (rt17603) [inline openghg/standardise/oned_timeseries/_crf.py:18]: Should this say parse "CRF" emissions data? (Similarly in description of `filepath`)

File: openghg/standardise/oned_timeseries/_crf.py
Comment (rt17603) [inline openghg/standardise/oned_timeseries/_crf.py:23]:
```suggestion
        source: Flux / Emissions source e.g. "energy"
```

File: openghg/store/_oneD_timeseries.py
Comment (rt17603) [inline openghg/store/_oneD_timeseries.py:334]: Would likely want to update e.g. something like "flux_timeseries"

File: openghg/standardise/oned_timeseries/_crf.py
Comment (rt17603) [inline openghg/standardise/oned_timeseries/_crf.py:63]: Based on `period` and `continuous` inputs can also include line to add `period_str` (and `start_date` and `end_date`) to metadata e.g. 
https://github.com/openghg/openghg/blob/devel/openghg/standardise/footprint/_acrg_org.py#L185C1-L187C6

```
start_date, end_date, period_str = infer_date_range(
        fp_time, filepath=input_filepath, period=period, continuous=continuous
    )

metadata["start_date"] = str(start_date)
metadata["end_date"] = str(end_date)
metadata["time_period"] = period_str
```

**Review by rt17603 — CHANGES_REQUESTED** [review]
Thanks for putting this together.

A few small points:
 - Just check over the documentation for "boundary conditions" and "intem" as I think this may have slipped through from where this was copied from.
 - The `source_format` keywords should be added to `standardise_*` and passed through. 
 - The `period` (and `continuous`) should also be added - we at least want `period`. `continuous` is in the process of being phased out but is still used for consistency.
 - For defining keys suggest using `region` (in preference to `domain`). Can still include `domain` as optional though for a specific meaning

Also a few other comments provided.

I think you mentioned this already but this does also need a test case.

File: openghg/store/_oneD_timeseries.py
Comment (SutarPrasad) [reply openghg/store/_oneD_timeseries.py:334]: updated to flux as we discussed in the meeting to be consistent.

File: openghg/standardise/oned_timeseries/_crf.py
Comment (SutarPrasad) [reply openghg/standardise/oned_timeseries/_crf.py:63]: (specific for CRF parser) start and end data can be updated by applying array indexing and stored in metadata, shall I just apply that?

**Comment by gareth-j** [comment]
Before I review this I think @rt17603's comments need to be resolved.

File: openghg/standardise/_standardise.py
Comment (SutarPrasad) [reply openghg/standardise/_standardise.py:886]: done

File: openghg/standardise/oned_timeseries/_crf.py
Comment (SutarPrasad) [reply openghg/standardise/oned_timeseries/_crf.py]: done

File: openghg/store/_oneD_timeseries.py
Comment (SutarPrasad) [reply openghg/store/_oneD_timeseries.py]: similar to flux implementation

**Comment by SutarPrasad** [comment]
> Before I review this I think @rt17603's comments need to be resolved.

Hi @gareth-j I have resolved those.

**Comment by rt17603** [comment]
> > Before I review this I think @rt17603's comments need to be resolved.
> 
> Hi @gareth-j I have resolved those.

Hi Prasad, it may be this just hasn't been pushed yet but I'm not currently seeing those changes reflected in the code?

**Comment by rt17603** [comment]
> > > Before I review this I think @rt17603's comments need to be resolved.
> > 
> > 
> > Hi @gareth-j I have resolved those.
> 
> Hi Prasad, it may be this just hasn't been pushed yet but I'm not currently seeing those changes reflected in the code?

Actually I can see some of them but I can still see references to "boundary conditions" for instance within the `standardise_flux_timeseries` doc string at the moment.

File: openghg/standardise/_standardise.py
Comment (rt17603) [reply openghg/standardise/_standardise.py:886]: Happy to discuss, but I don't see the `region` keyword added here if that's what we decided was the best approach?

**Comment by SutarPrasad** [comment]
> > > > Before I review this I think @rt17603's comments need to be resolved.
> > > 
> > > 
> > > Hi @gareth-j I have resolved those.
> > 
> > 
> > Hi Prasad, it may be this just hasn't been pushed yet but I'm not currently seeing those changes reflected in the code?
> 
> Actually I can see some of them but I can still see references to "boundary conditions" for instance within the `standardise_flux_timeseries` doc string at the moment.

Oh, I see the confusion my bad, I had this branch open on local as well Blue Pebble, but the changes I made I did not sync up correctly. Apologies for this, I will sync vscode instance with this.

File: openghg/standardise/flux_timeseries/_crf.py
Comment (rt17603) [inline openghg/standardise/flux_timeseries/_crf.py:10]: The `source` input is included to describe where the emissions are coming from. In the case of the CRF data this may be best described as anthropogenic so a more intuitive default could be something like: "anthro".

File: openghg/store/_flux_timeseries.py
Comment (rt17603) [inline openghg/store/_flux_timeseries.py:248]: @brendan-m-murphy - can you recall the discussion on this? For the 1D flux timeseries with (time,) dimension would we want this to have the same variable naming as for the 3D flux data with (time, lat, lon). We can do so but this could potentially cause confusion.

File: openghg/store/_oneD_timeseries.py
Comment (rt17603) [reply openghg/store/_oneD_timeseries.py]: @brendan-m-murphy - can you recall the discussion on this? For the 1D flux timeseries with (time,) dimension would we want this to have the same variable naming as for the 3D flux data with (time, lat, lon). We can do so but this could potentially cause confusion.

File: openghg/standardise/flux_timeseries/_crf.py
Comment (SutarPrasad) [reply openghg/standardise/flux_timeseries/_crf.py]: updated to anthro

File: openghg/store/_oneD_timeseries.py
Comment (SutarPrasad) [reply openghg/store/_oneD_timeseries.py]: Waiting for this, Apart from that I believe everything is resolved.

File: openghg/store/base/_datasource.py
Comment (gareth-j) [inline openghg/store/base/_datasource.py:130]:
```suggestion
            data_type: Type of data, one of ["boundary_conditions", "column", "emissions", "flux", "flux_timeseries", "footprints", "surface", "eulerian_model"].
```

**Review by gareth-j — CHANGES_REQUESTED** [review]
Just a few small changes needed.

**Comment by SutarPrasad** [comment]
> Just a few small changes needed.

Done

**Comment by gareth-j** [comment]
We just need a section in the docs added for adding CRF data and then this will be ready.

File: openghg/util/_util.py
Comment (gareth-j) [inline openghg/util/_util.py:135]:
```suggestion
    matches: List = process.extract(site_name, site_list)
```

**Review by gareth-j — DISMISSED** [review]
Looks good!

File: openghg/store/_oneD_timeseries.py
Comment (rt17603) [reply openghg/store/_oneD_timeseries.py]: Let's update the variable name to `flux_timeseries` (1D) for now to keep this distinct from `flux` (3D).

**Comment by SutarPrasad** [comment]
Hi,

As per suggestion updated “flux” to “flux_timeseries”

Thanks,
Prasad

From: Rachel Tunnicliffe ***@***.***>
Date: Thursday, 9 May 2024 at 14:15
To: openghg/openghg ***@***.***>
Cc: Prasad Sutar ***@***.***>, Author ***@***.***>
Subject: Re: [openghg/openghg] Class and parse_crf for 1D data (PR #870)

@rt17603 commented on this pull request.

________________________________

In openghg/store/_oneD_timeseries.py<https://github.com/openghg/openghg/pull/870#discussion_r1595430825>:

> +

+        Includes observation for each time of the defined domain:

+            - "Obs"

+                - expected dimensions: ("time")

+

+        Expected data types for all variables and coordinates also included.

+

+        Returns:

+            DataSchema : Contains schema for OneDTimeseries.

+        """

+        from openghg.store import DataSchema

+

+        data_vars: Dict[str, Tuple[str, ...]] = {"oneD": ("time",)}

+        dtypes = {

+            "time": np.datetime64,

+            "oneD": np.floating,

Let's update the variable name to flux_timeseries (1D) for now to keep this distinct from flux (3D).

File: openghg/standardise/_standardise.py
Comment (rt17603) [inline openghg/standardise/_standardise.py:885]: Picky comment but I think the keywords should be rearranged here to emphasise importance:
```
    region: str = "UK",
    source_format: str = "crf",
    domain: Optional[str] = None,
```

For now we can leave the defaults of "crf" and "UK" but if we add any other flux_timeseries parser functions we should also remove these defaults.

I think if we have `domain` above `region` this will get in preference and in circumstances when it doesn't make the most sense.

File: openghg/standardise/_standardise.py
Comment (rt17603) [inline openghg/standardise/_standardise.py:906]:
```suggestion
        domain: If flux is related to pre-existing domain (e.g. "EUROPE") with defined latitude-longitude bounds this can be used to flag that. Otherwise, use `region` input to describe the name of a region (e.g. "UK").
```
As per the suggestion on argument order, the arguments should be reordered in the doc-string in the same way.

File: openghg/standardise/flux_timeseries/_crf.py
Comment (rt17603) [inline openghg/standardise/flux_timeseries/_crf.py:16]: Similiarly, would be good to make sure `region` is above `domain` in the function call to emphasise it's use as an identifying key.

File: openghg/store/_flux_timeseries.py
Comment (rt17603) [inline openghg/store/_flux_timeseries.py:124]: The `region` keyword should also be passed to the `clean_string` function for consistency.

File: tests/standardise/test_standardise.py
Comment (rt17603) [inline tests/standardise/test_standardise.py:475]: Just a note that if the `region` keyword gets passed to the `clean_string` function (as mentioned above), the key here will include a lowercase "UK" e.g.
```suggestion
    assert "ch4_crf_uk" in flux_results
```

**Review by rt17603 — CHANGES_REQUESTED** [review]
A few very small comments around the `region` keyword to make sure this is emphasised and formatted but otherwise looks good.

File: openghg/store/_flux_timeseries.py
Comment (rt17603) [inline openghg/store/_flux_timeseries.py:197]: Also just noticed, we'd likely need the required keys to also include `region` as we will always need to define which area the flux is related to when storing this (c.f. `domain` for `flux` inputs).

**Review by rt17603 — CHANGES_REQUESTED** [review]
One small change (just `domain` back to `region` on one line) then ready to go.

**Comment by SutarPrasad** [comment]
It's ready now, the failed job has passed.

**Comment by rt17603** [comment]
This is ready to go but we should wait to merge this until after the v0.8.1 release is completed. As soon as v0.8.1 out we should approve and merge this.

**Review by gareth-j — APPROVED** [review]
Looks good!

**Review by rt17603 — APPROVED** [review]

**Pull request merged**
